src/plugin/sdl/sdl.c

```diff
@@ -1214,25 +1215,34 @@ static void SDL_draw_string(void *opaque, int x, int y, unsigned char *text, int
                                            text_colors[ATTR_BG(attr)]);
   pthread_mutex_unlock(&sdl_font_mtx);
   free(s);
-  SDL_Texture *txt = SDL_CreateTextureFromSurface(renderer, srf);
 
   SDL_Rect rect;
   rect.x = font_width * x;
   rect.y = font_height * y; /* height plus spacing to next line */
   rect.w = srf->w;
   rect.h = srf->h;
 
+  pthread_mutex_lock(&rend_mtx);
+#if 0 // currently not working
```

Comment:
OK, forget about BlitSurface then.
See SDL_put_image() and just do the same
to update `texture_buf`.
Unfortunately `texture_buf` is also not
created for text modes currently, so all
the same changes are still needed...

Maybe I also do not understand your
idea fully. I see you do RenderCopy(),
but where is RenderPresent()?

As for rectangles...
https://wiki.libsdl.org/MigrationGuide
It says:
```
This will upload your pixels to GPU memory. That NULL can be a subregion if you want to mess around with dirty
rectangles, but chances are modern hardware can just swallow the whole framebuffer without much trouble.
```
But they apply it to `SDL_UpdateTexture()`
not RenderCopy()! So I am not even sure if
their statement is valid, or my understanding
is.
So there is a big risk here that I can misguide
you. Anyway, I still don't see where you do
RenderPresent().

Reply:
Perhaps they do not mean rectangles are
not recommended. Maybe they are just saying
that modern hw can easily handle full framebuffer,
so the one may not even care about rectangles?
In that case I simply misread them, and there is
no recommendation against rectangles at all.

Reply:
And by "dirty rectangles" they obviously
meant "updated rectangles" rather than
something not recommended... So yes, I
misread this all. :)

Reply:
>  I see you do RenderCopy(),
but where is RenderPresent()?

I thought that an alternative was incrementing `sdl_rects` and the `RenderPresent()` was done sometime later (perhaps by another thread?)? As I test I added the `RenderPresent()` call and didn't increment `sdl_rects` but it didn't help with the 'tearing' and it does appear a little slower.

I'm sure your idea for `BlitSurface()` is a good one, but I think we have to setup the surface properly for text mode. Commenting out the `SDL_change_mode(0,0` allowed the surface to exist, but I'm sure its geometry was incorrect.

But as I'm sure you can tell this is my first time messing with SDL, so really I have no proper idea.

Reply:
OK, I am confused.
When we use $_X_font, we just switch
to X. SDL is then not used. So I wonder
why do we have all that X font code in
SDL? Maybe its not needed eve now?
Can you split out the removal patch
and provide it in another PR?

> I thought that an alternative was incrementing sdl_rects and the RenderPresent()
> was done sometime later (perhaps by another thread?)?

Maybe, but I have hard times seeing how...
See `do_rend()`. It does `SDL_RenderClear()`
first, so all your rendering would just be lost.
Next it does `SDL_RenderCopy(renderer, texture_buf, NULL, NULL);`
and `texture_buf` is not created for text modes,
together with the main surface. So that would
just SIGSEGV. Which is why I think the current
SDL code is not suitable for X fonts. The fact
that the X fonts code is there, and the fact that
you got it to work w/o forcibly creating `texture_buf`
makes me very confused.

Reply:
> Can you split out the removal patch
and provide it in another PR?

But if I do that without adding the ttf support, we won't be able to draw text, except in bitmap mode?

Reply:
No, and that's the whole point: we currently
fall back to an X plugin completely for $_X_font.
And do_rend() of SDL would just SIGSEGV
if it is called when texture_buf is not created.
So it shouldn't be called in this case, and yet
you say it does, because someone does the
rendering for you?
Anyway, since we fall back to an X plugin, I
don't understand where that X supporting
code in SDL came from and did it ever work...

Reply:
Okay, I'll rip it out and see what happens...

Reply:
So I pushed a new branch [sdltxt-01](https://github.com/andrewbird/dosemu2/tree/sdltxt-01) with the X support removed from sdl.c. So it really does fall back to use the whole X plugin and not just using bits of it in the SDL plugin!

Reply:
Yeah, that's quite surprising.
But fewer work to you. :)

Reply:
I was also going to remove the unnecessary function prototypes from X.h and make them static, but I see sdl1 plugin still uses them. So I guess for now they get to stay?

Reply:
Yes, don't spend too much time on legacy.
Hope we will forget about X plugin soon.
What is interesting is how the rendering is
done in your code if there is no texture_buf,
so if there any rendering thread - it should
just crash...